core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -142,18 +142,27 @@ private static boolean isMasterChannelException(TransportException exp) {
      */
     public void remoteShardFailed(final ShardId shardId, String allocationId, long primaryTerm, final String message, @Nullable final Exception failure, Listener listener) {
         assert primaryTerm > 0L : "primary term should be strictly positive";
-        shardFailed(shardId, allocationId, primaryTerm, message, failure, listener);
+        shardFailed(shardId, allocationId, primaryTerm, message, failure, listener, clusterService.state());
     }
 
     /**
      * Send a shard failed request to the master node to update the cluster state when a shard on the local node failed.
      */
     public void localShardFailed(final ShardRouting shardRouting, final String message, @Nullable final Exception failure, Listener listener) {
-        shardFailed(shardRouting.shardId(), shardRouting.allocationId().getId(), 0L, message, failure, listener);
+        localShardFailed(shardRouting, message, failure, listener, clusterService.state());
     }
 
-    private void shardFailed(final ShardId shardId, String allocationId, long primaryTerm, final String message, @Nullable final Exception failure, Listener listener) {
-        ClusterStateObserver observer = new ClusterStateObserver(clusterService, null, logger, threadPool.getThreadContext());
+    /**
+     * Send a shard failed request to the master node to update the cluster state when a shard on the local node failed.
+     */
+    public void localShardFailed(final ShardRouting shardRouting, final String message, @Nullable final Exception failure, Listener listener,
+                                 final ClusterState currentState) {
```

Comment:
I wonder if we need to change all the methods to pass the ClusterState down here. We care about the master node here so that we know where to send the shard failed message to. This is a more frequent pattern that is also used in SnapshotService so maybe it makes sense to build some extra infra for this (e.g. through a MasterNotificationService that is also a StateApplier but updated first).
Does not need to be addressed in this PR, but just wanted to hear your thoughts on this.